msos/WaitChains.cs

```diff
@@ -17,68 +18,561 @@ class WaitChains : ICommand
                 String.Format("~ {0}; !clrstack", thread.ManagedThreadId));
             _context.WriteLine();
-            
+
             if (visitedThreadIds.Contains(thread.ManagedThreadId))
             {
                 _context.WriteLine("{0}*** DEADLOCK!", new string(' ', depth * 2));
                 return;
             }
             visitedThreadIds.Add(thread.ManagedThreadId);
 
-            foreach (var blockingObject in thread.BlockingObjects)
+            if (unified_thread.BlockingObjects != null)
             {
-                _context.Write("{0}| {1} ", new string(' ', (depth+1) * 2), blockingObject.Reason);
-                var type = _context.Heap.GetObjectType(blockingObject.Object);
-                if (type != null && !String.IsNullOrEmpty(type.Name))
+                foreach (var blockingObject in unified_thread.BlockingObjects)
                 {
-                    _context.WriteLink(
-                        String.Format("{0:x16} {1}", blockingObject.Object, type.Name),
-                        String.Format("!do {0:x16}", blockingObject.Object));
+                    
+                    _context.Write("{0}| {1}", new string(' ', (depth + 1) * 2), blockingObject.Reason);
+
+                    if (!String.IsNullOrEmpty(blockingObject.KernelObjectName))
+                    {
+                        _context.WriteLink(
+                            String.Format("{0:x16} {1} {2}", blockingObject.Handle,
+                            blockingObject.KernelObjectTypeName, blockingObject.KernelObjectName),
+                            String.Format("!do {0:x16}", blockingObject.Handle));
+                    }
+
+                    var type = _context.Heap.GetObjectType(blockingObject.ManagedObjectAddress);
+                    if (type != null && !String.IsNullOrEmpty(type.Name))
+                    {
+                        _context.WriteLink(
+                            String.Format("{0:x16} {1}", blockingObject.ManagedObjectAddress, type.Name),
+                            String.Format("!do {0:x16}", blockingObject.ManagedObjectAddress));
+                    }
+                    else
+                    {
+                        _context.Write("{0:x16}", blockingObject.ManagedObjectAddress);
+                    }
+
+                    _context.WriteLine();
+
+                    if (blockingObject.Owners != null)
+                    {
+                        foreach (var owner in blockingObject.Owners)
+                        {
+                            if (owner == null) // ClrMD sometimes reports this nonsense
+                                continue;
+
+                            DisplayChainForThread(owner, depth + 2, visitedThreadIds);
+                        }
+                    }
                 }
-                else
+            }
+        }
+
+        private void DisplayUnManagedThread_ChainAux(UnifiedThread thread, int depth, HashSet<int> visitedThreadIds)
```

Comment:
This method and the previous one look almost the same, and contain a lot of code duplication. Please try to minimize the duplication.